### PR TITLE
mount kernel filesystems if run as pid1

### PR DESCRIPTION
**Pull request by Vincinator**
Hi aurae community,

This PR adds first steps to make auraed boot as pid1, mainly mounting the kernel filesystems (devfs, procfs and sysfs), but also some minor fixes to make auraed boot without kernel panic.

* Fixes kernel panic due to a rust panic thrown by syslog creation (missing unix sockets [1])
    * The rust panic causes auraed to stop, and if started as pid1 the kernel will in turn also panic because the init process ended.
* Fixes `hack/hack.mk` for `zsh` users
* Adds a SystemRuntime (analog to AuraedRuntime), which handles system preparation. 
    * SystemRuntime Case: not-pid1
        * assume a systemd environment (currently only unix socket `/var/run/syslog`)
    * SystemRuntime Case: pid1 
        * initializes logging without unix socket dependency. Using syslog crate would require creation of unix socket first, and I suggest to discuss this first if there is a use case to have a unix socket for logging.
        * mounts file based kernel interfaces
      * devfs, sysfs, procfs
* Added helper functions for basic file io (currently for devfs, sysfs, procfs interfaces)
    * get content of e.g. `/dev`, `/sys` or `/proc`
    * or `/sys/class/net/` to see available network interfaces

## ToDo: 
* Discuss Logging targets
    * Same log targets for all SystemRuntime cases? 
* Network initialization
* Stop auraed in pid1 case gracefully (kernel halt syscall to shutdown or reboot)

## How to test

Steps to run auraed as pid in a qemu vm are documented in the [hack/README.md](https://github.com/aurae-runtime/auraed/blob/main/hack/README.md). Copy pasted them here:

    make build-container
    make kernel
    make initramfs

    # create `vm-br0` bridge on your machine:
    make network

    # run auraed in a VM as pid 1:
    make virsh-start virsh-console virsh-stop

    # exit VM console with Ctrl+]

Screenshot 2022-09-27 at 17 01 29

## Links
[1] https://docs.rs/syslog/latest/src/syslog/lib.rs.html#232-243

### CODE REVIEW

**Comment by MalteJ** [comment]
@cla-bot check

**Comment by Vincinator** [comment]
@cla-bot check

**Comment by MalteJ** [comment]
@cla-bot check

File: src/system/init.rs [hunk 1]
```diff
@@ -0,0 +1,93 @@
+use log::*;
```
Comment (krisnova) [inline src/system/init.rs]: Can we add the header please?

File: src/system/init.rs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    ) {
+        Ok(_) => {}
+        Err(e) => panic!("unable to connect to syslog: {:?}", e),
+    };
+}
+
+// To Discuss (TODO):
+//      The "syslog" logger requires unix sockets.
+//      Syslog assumes that either /dev/log or /var/run/syslog are available [1].
+//      We need to discuss if there is a use case to log via unix sockets,
```
Comment (krisnova) [inline src/system/init.rs]: In general the project should support the ability to log to a unix socket, however this is a fine starting point for right now. 

Opened https://github.com/aurae-runtime/auraed/issues/24 to track the work.

File: src/system/mod.rs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+pub(crate) mod fileio;
+pub(crate) mod init;
+
+const AURAED_SYSLOG_NAME: &str = "auraed";
+
+pub fn get_pid() -> u32 {
+    std::process::id()
+}
+
+pub fn print_logo() {
```
Comment (krisnova) [inline src/system/mod.rs]: So I love the logo! I already tweeted it! https://twitter.com/krisnova/status/1575274332665815044

Can we break this into 2 functions?

```rust
pub fn banner() -> String {
    format!("...")
}
```

and a 2nd function that prints the output of the first?

```rust
pub fn print_logo() -> String {
    println!("{}", banner())
}
```

I plan on using the banner in other places in the program eventually :)

File: src/system/init.rs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+pub fn init_rootfs() {
+    if get_pid() != 1 {
+        warn!("Trying to initialize rootfs but auraed is not run as pid1. Abort setup of rootfs.");
+        return;
+    }
+
+    mount_vfs("none", "/dev", "devtmpfs");
+    mount_vfs("none", "/sys", "sysfs");
+    mount_vfs("proc", "/proc", "proc");
```
Comment (krisnova) [inline src/system/init.rs]: This sounds a bit crazy, however I have a use case for a system where `procfs` is NOT mounted.

Is it possible to add a compile time flag or similar to run `auraed` in an extremely risky experimental mode where `procfs` is not mounted?

File: src/system/fileio.rs [hunk 5]
```diff
@@ -0,0 +1,31 @@
+use std::fs;
```
Comment (krisnova) [inline src/system/fileio.rs]: Header

**Review by krisnova — APPROVED** [review]
My suggestion is to go ahead and merge this whenever you feel confident @Vincinator 

You should be able to merge now. I will rebase #17 on top of your work, and I encourage you to iterate as you see best. If you would like to pick up more features in subsequent PRs please continue to work however is best for you.

**Comment by krisnova** [comment]
Non blocker however I am wondering about the name `system` for a module for init. This seems to be overloaded. I am unsure how I feel about it. Does anyone else have any thoughts?

**Comment by Vincinator** [comment]
> Non blocker however I am wondering about the name `system` for a module for init. This seems to be overloaded. I am unsure how I feel about it. Does anyone else have any thoughts?

My vote goes to naming it `init`. Current scope is just initialization of stuff for the user space.

File: src/system/init.rs (hunk 4 above)
Comment (Vincinator) [reply src/system/init.rs]: Sure! Will add a simple and fast solution to this PR. Maybe it makes sense to have the mount points configurable? Opened an issue here to discuss and track https://github.com/aurae-runtime/auraed/issues/25

Comment (MalteJ) [reply src/system/init.rs]: Maybe a `noprocfs` kernel parameter?

Comment (Vincinator) [reply src/system/init.rs]: Ok, I was too optimistic that I can add this fast and simple. Added my comment to the discussion on discord. 

I think this requires a second PR

**Comment by Vincinator** [comment]
@kris-nova this PR is ready to be merged from my point of view 
Thanks for your feedback!

**Review by MalteJ — APPROVED** [review]

**Comment by krisnova** [comment]
Note for myself. Configuration to remove `/proc` procfs(5) is in this issue:  https://github.com/aurae-runtime/auraed/issues/25

**Pull request merged**